torch/csrc/jit/pickler.h

```diff
@@ -0,0 +1,181 @@
+#include <string>
+#include <vector>
+
+#include <ATen/core/ivalue.h>
+
+namespace torch {
+namespace jit {
+
+// See Python's pickletools.py for a detailed description of each of these codes
+enum class OpCode : char {
+  MARK = '(',
+  STOP = '.',
+  POP = '0',
+  POP_MARK = '1',
+  DUP = '2',
+  FLOAT = 'F',
+  INT = 'I',
+  BININT = 'J',
+  BININT1 = 'K',
+  LONG = 'L',
+  BININT2 = 'M',
+  NONE = 'N',
+  PERSID = 'P',
+  BINPERSID = 'Q',
+  REDUCE = 'R',
+  STRING = 'S',
+  BINSTRING = 'T',
+  SHORT_BINSTRING = 'U',
+  UNICODE = 'V',
+  BINUNICODE = 'X',
+  APPEND = 'a',
+  BUILD = 'b',
+  GLOBAL = 'c',
+  DICT = 'd',
+  EMPTY_DICT = '}',
+  APPENDS = 'e',
+  GET = 'g',
+  BINGET = 'h',
+  INST = 'i',
+  LONG_BINGET = 'j',
+  LIST = 'l',
+  EMPTY_LIST = ']',
+  OBJ = 'o',
+  PUT = 'p',
+  BINPUT = 'q',
+  LONG_BINPUT = 'r',
+  SETITEM = 's',
+  TUPLE = 't',
+  EMPTY_TUPLE = ')',
+  SETITEMS = 'u',
+  BINFLOAT = 'G',
+
+  // Protocol 2
+  PROTO = '\x80',
+  NEWOBJ = '\x81',
+  EXT1 = '\x82',
+  EXT2 = '\x83',
+  EXT4 = '\x84',
+  TUPLE1 = '\x85',
+  TUPLE2 = '\x86',
+  TUPLE3 = '\x87',
+  NEWTRUE = '\x88',
+  NEWFALSE = '\x89',
+  LONG1 = '\x8a',
+  LONG4 = '\x8b',
+
+  // Protocol 3 (Python 3.x)
+  BINBYTES = 'B',
+  SHORT_BINBYTES = 'C',
+
+  // Protocol 4
+  SHORT_BINUNICODE = '\x8c',
+  BINUNICODE8 = '\x8d',
+  BINBYTES8 = '\x8e',
+  EMPTY_SET = '\x8f',
+  ADDITEMS = '\x90',
+  FROZENSET = '\x91',
+  NEWOBJ_EX = '\x92',
+  STACK_GLOBAL = '\x93',
+  MEMOIZE = '\x94',
+  FRAME = '\x95'
+};
+
+using ::c10::IValue;
+
+struct Pickler {
+  Pickler(std::vector<at::Tensor>& tensor_table)
+      : tensor_table_(tensor_table),
+        tensor_class_name_("__main__\n", "TensorID\n") {}
+
+  const std::vector<char>& stack();
+  void start();
+  void finish();
+  void addIValue(const IValue& ivalue);
+
+ private:
+  void pushBinGet(uint32_t memo_id);
+  void pushString(const IValue& ivalue);
+  void pushString(const std::string& string);
+  void pushTensor(const IValue& ivalue);
+  void pushDouble(const IValue& ivalue);
+  void pushMemoization(const void* item);
+  void pushMemoization(const IValue& ivalue);
+  void pushList(const IValue& ivalue);
+  void pushTuple(const IValue& ivalue);
+  void pushDict(const IValue& ivalue);
+  const void* getPointer(const IValue& ivalue);
+
+  template <typename T>
```

Comment:
I recommend just hard coding pushUInt8, pushOpCode, and pushUInt32. Those are the only ones used. This comment and the class is longer than just writing these three functions out and more confusing!